# Clean Data and Analyze the Used Car Listings on eBay in Germany

Dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay websit.
The data had been modified by dataquest for teaching purpose.
Modifications made:
    
* We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
* We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)


Data dictionary:

| columns | description |
|-----:|:----|
| dateCrawled | When this ad was first crawled. All field-values are taken from this date.|
| name | Name of the car.|
| seller | Whether the seller is private or a dealer.|
| offerType | The type of listing|
| price | The price on the ad to sell the car.|
| abtest | Whether the listing is included in an A/B test.|
| vehicleType | The vehicle Type.|
| yearOfRegistration | The year in which the car was first registered.|
| gearbox | The transmission type.|
| powerPS | The power of the car in PS.|
| model | The car model name.|
| kilometer | How many kilometers the car has driven.|
| monthOfRegistration | The month in which the car was first registered.|
| fuelType | What type of fuel the car uses.|
| brand | The brand of the car.|
| notRepairedDamage | If the car has a damage which is not yet repaired.|
| dateCreated | The date on which the eBay listing was created.|
| nrOfPictures | The number of pictures in the ad.|
| postalCode | The postal code for the location of the vehicle.|
| lastSeenOnline | When the crawler saw this ad last online.|


In [1]:
# Import libraries`
import pandas as pd
import numpy as np

In [2]:
# Read in data as a pandas object
# Use Latin-1 as encoding as there are special chacters in the file
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
# Display the first 5 lines of the data
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Observations:
1. There are two types of data: String (time as String type) and Integer
2. Columns price, odometer are mixed with charaters and numbers
3. Column names are mixed with big and little letters, with some in camelCase

In [4]:
# Get a concise summary of the dataframe
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

Observation:
1. There are two types of data types: integer (5 columns) and string(15 columns)
2. vehicleType, gearbox, model, fuelType and notRepairedDamage have missing values as there number of entries (for example vehicleTyep have 44,905 entries) are smaller than total number of entries (50,000)

### Modify column names
* Convert from camelCase to snake_case
* Reword some of the column names based on the data dictionary to be more descriptive

In [5]:
# print the existing column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

**Change column names**

In [6]:
# create new column names
new_column_names = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

# assign new column names to the dataframe
autos.columns = new_column_names

# display the new column names
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [7]:
# Show the first few lines to see the changes
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

**We look for the text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.**

In [8]:
# get descriptive statistics for all columns
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Observation:
* Columns that have mostly one value that are candiates to be dropped: seller, offer_type, abtest, gearbox, unrepaired_damage. We keep those columns for now.
* Numeric data stored as text that needs to be cleaned: price, odometer

In [9]:
# Remove non-numeric characters from columns price and odometer
autos["price"] = autos["price"].str.replace("$","").str.replace(",","")
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","")

In [10]:
# Convert price and odometer columns to a numeric dtype
autos['price'] = autos['price'].astype(int)
autos["odometer"] = autos["odometer"].astype(int)

In [11]:
# rename column odometer name
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

**Exploring data: looking for data that might be wrong**

In [12]:
autos['price'].unique().shape

(2357,)

In [13]:
# basic statistics of the price column
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Min price is 0, and max price is 100,000,000. Those are the outliers.

In [14]:
# Count frequences of each price number, sort and show the first 20
autos['price'].value_counts().sort_index().head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

**Count frequences of each price number**

In [15]:
# Count frequences of each price number, sort and whow the last 20
autos['price'].value_counts().sort_index().tail(20)

197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
1234566     1
1300000     1
3890000     1
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price, dtype: int64

**Display highest price**

In [16]:
# display highest price: 
# sort the counts in descending order and then show the first 20
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

**Display lowest price**

In [17]:
# display lowest price:
# sort the counts in ascending order and then show the first few
autos['price'].value_counts().sort_index(ascending=True).head()

0    1421
1     156
2       3
3       1
5       2
Name: price, dtype: int64

**Display highest odometer**

In [18]:
# display highest odometer
autos['odometer_km'].value_counts().sort_index(ascending=False).head(20)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

**Display lowest odometer**

In [19]:
# display lowest odometer
autos['odometer_km'].value_counts().sort_index(ascending=True).head(20)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

**Remove outliers based on price**

In [21]:
# remove outliers based on price, 
# keep the entries with prices between 1,000 and 5,000 
# as these prices are close to reality
autos = autos[autos["price"].between(1000,5000)]

# display descriptive statistics for all columns
# after removing outliers
autos.describe(include='all')

date_crawled      name  seller offer_type         price abtest  \
count                 21632     21632   21632      21632  21632.000000  21632   
unique                21299     17009       1          1           NaN      2   
top     2016-03-08 10:40:35  BMW_316i  privat    Angebot           NaN   test   
freq                      3        55   21632      21632           NaN  11241   
mean                    NaN       NaN     NaN        NaN   2603.944804    NaN   
std                     NaN       NaN     NaN        NaN   1161.029714    NaN   
min                     NaN       NaN     NaN        NaN   1000.000000    NaN   
25%                     NaN       NaN     NaN        NaN   1579.750000    NaN   
50%                     NaN       NaN     NaN        NaN   2450.000000    NaN   
75%                     NaN       NaN     NaN        NaN   3500.000000    NaN   
max                     NaN       NaN     NaN        NaN   5000.000000    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         19622        21632.00000    20566  21632.000000  20611   
unique            8                NaN        2           NaN    213   
top       limousine                NaN  manuell           NaN   golf   
freq           5785                NaN    17237           NaN   1816   
mean            NaN         2002.52242      NaN    102.949935    NaN   
std             NaN           27.12203      NaN    174.232012    NaN   
min             NaN         1000.00000      NaN      0.000000    NaN   
25%             NaN         1999.00000      NaN     69.000000    NaN   
50%             NaN         2002.00000      NaN    101.000000    NaN   
75%             NaN         2005.00000      NaN    131.000000    NaN   
max             NaN         5000.00000      NaN  16011.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    21632.000000        21632.000000     19900       21632   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     13741        4657   
mean    135861.917530            5.829558       NaN         NaN   
std      29984.319749            3.676373       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              17436                21632         21632.0  21632.000000   
unique                 2                   54             NaN           NaN   
top                 nein  2016-03-12 00:00:00             NaN           NaN   
freq               15547                  830             NaN           NaN   
mean                 NaN                  NaN             0.0  50227.746810   
std                  NaN                  NaN             0.0  25521.835303   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  29571.750000   
50%                  NaN                  NaN             0.0  48691.000000   
75%                  NaN                  NaN             0.0  70376.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 21632  
unique                19638  
top     2016-04-06 15:16:45  
freq                      5  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### Distribution of dates related to cars listed

In [23]:
# distribution of date crawled
dates_crawled = autos["date_crawled"].str[:10]

# use normalize=True to use percentage instead of counts
# use dropna=False to include missing values in the distribution
dates_crawled.value_counts(normalize=True, dropna=False)

2016-03-12    0.038462
2016-03-21    0.038230
2016-04-03    0.037445
2016-03-14    0.036705
2016-03-20    0.036566
2016-04-04    0.036566
2016-03-07    0.036520
2016-03-28    0.035318
2016-04-02    0.035133
2016-03-19    0.034994
2016-03-08    0.034486
2016-03-10    0.034393
2016-03-15    0.033793
2016-03-29    0.033515
2016-03-22    0.033515
2016-03-09    0.033330
2016-03-11    0.033145
2016-03-30    0.033099
2016-03-26    0.032776
2016-04-01    0.032359
2016-03-31    0.031666
2016-03-17    0.031481
2016-03-23    0.031435
2016-03-27    0.030973
2016-03-25    0.030233
2016-03-16    0.029447
2016-03-24    0.029401
2016-03-05    0.025287
2016-03-13    0.014469
2016-03-06    0.014007
2016-04-05    0.013406
2016-03-18    0.012574
2016-04-06    0.003883
2016-04-07    0.001387
Name: date_crawled, dtype: float64

**It's not easy to summerize the distriution, sort it by dates:**

In [32]:
dates_crawled.value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025287
2016-03-06    0.014007
2016-03-07    0.036520
2016-03-08    0.034486
2016-03-09    0.033330
2016-03-10    0.034393
2016-03-11    0.033145
2016-03-12    0.038462
2016-03-13    0.014469
2016-03-14    0.036705
2016-03-15    0.033793
2016-03-16    0.029447
2016-03-17    0.031481
2016-03-18    0.012574
2016-03-19    0.034994
2016-03-20    0.036566
2016-03-21    0.038230
2016-03-22    0.033515
2016-03-23    0.031435
2016-03-24    0.029401
2016-03-25    0.030233
2016-03-26    0.032776
2016-03-27    0.030973
2016-03-28    0.035318
2016-03-29    0.033515
2016-03-30    0.033099
2016-03-31    0.031666
2016-04-01    0.032359
2016-04-02    0.035133
2016-04-03    0.037445
2016-04-04    0.036566
2016-04-05    0.013406
2016-04-06    0.003883
2016-04-07    0.001387
Name: date_crawled, dtype: float64

We can see that on Mar 6 (Sunday), 13(Sunday) and 18(Friday), percentage of crawelling are low, maybe that's because these are weekend. It also shows that at the beginning of months, celling are low. At other dates, the selling are generally equal.

**Sort it by percentage of crawelling**

In [31]:
dates_crawled.value_counts(normalize=True, dropna=False).sort_values()

2016-04-07    0.001387
2016-04-06    0.003883
2016-03-18    0.012574
2016-04-05    0.013406
2016-03-06    0.014007
2016-03-13    0.014469
2016-03-05    0.025287
2016-03-24    0.029401
2016-03-16    0.029447
2016-03-25    0.030233
2016-03-27    0.030973
2016-03-23    0.031435
2016-03-17    0.031481
2016-03-31    0.031666
2016-04-01    0.032359
2016-03-26    0.032776
2016-03-30    0.033099
2016-03-11    0.033145
2016-03-09    0.033330
2016-03-22    0.033515
2016-03-29    0.033515
2016-03-15    0.033793
2016-03-10    0.034393
2016-03-08    0.034486
2016-03-19    0.034994
2016-04-02    0.035133
2016-03-28    0.035318
2016-03-07    0.036520
2016-04-04    0.036566
2016-03-20    0.036566
2016-03-14    0.036705
2016-04-03    0.037445
2016-03-21    0.038230
2016-03-12    0.038462
Name: date_crawled, dtype: float64

The lowest crawelling is 0.01387% on 2016-04-07, while the highest is 3.8462% on 2016-03-12.

**Dates on which advertisement posted on eBay**

In [35]:
# Distribution of date on which eBay listing created
# sort by dates
dates_ad_created = autos["ad_created"].str[:10]
dates_ad_created.value_counts(normalize=True, dropna=False).sort_index()

2015-12-30    0.000046
2016-01-07    0.000046
2016-01-13    0.000046
2016-01-27    0.000092
2016-02-01    0.000046
2016-02-02    0.000046
2016-02-11    0.000046
2016-02-12    0.000046
2016-02-16    0.000046
2016-02-18    0.000046
2016-02-19    0.000046
2016-02-20    0.000046
2016-02-21    0.000046
2016-02-27    0.000139
2016-02-28    0.000139
2016-02-29    0.000185
2016-03-01    0.000139
2016-03-02    0.000092
2016-03-03    0.000971
2016-03-04    0.001618
2016-03-05    0.023068
2016-03-06    0.015348
2016-03-07    0.035272
2016-03-08    0.034209
2016-03-09    0.033746
2016-03-10    0.033931
2016-03-11    0.033145
2016-03-12    0.038369
2016-03-13    0.015995
2016-03-14    0.035226
2016-03-15    0.033654
2016-03-16    0.029956
2016-03-17    0.031204
2016-03-18    0.013267
2016-03-19    0.033746
2016-03-20    0.036844
2016-03-21    0.038323
2016-03-22    0.033469
2016-03-23    0.031481
2016-03-24    0.029077
2016-03-25    0.030510
2016-03-26    0.033007
2016-03-27    0.030510
2016-03-28 

In [36]:
# sort by percentage
dates_ad_created.value_counts(normalize=True, dropna=False).sort_values()

2016-01-07    0.000046
2016-02-19    0.000046
2016-02-20    0.000046
2016-02-01    0.000046
2015-12-30    0.000046
2016-02-12    0.000046
2016-02-21    0.000046
2016-02-02    0.000046
2016-02-11    0.000046
2016-02-16    0.000046
2016-02-18    0.000046
2016-01-13    0.000046
2016-01-27    0.000092
2016-03-02    0.000092
2016-02-27    0.000139
2016-02-28    0.000139
2016-03-01    0.000139
2016-02-29    0.000185
2016-03-03    0.000971
2016-04-07    0.001294
2016-03-04    0.001618
2016-04-06    0.003883
2016-04-05    0.011973
2016-03-18    0.013267
2016-03-06    0.015348
2016-03-13    0.015995
2016-03-05    0.023068
2016-03-24    0.029077
2016-03-16    0.029956
2016-03-25    0.030510
2016-03-27    0.030510
2016-03-17    0.031204
2016-03-23    0.031481
2016-03-31    0.031620
2016-04-01    0.032498
2016-03-30    0.032868
2016-03-26    0.033007
2016-03-11    0.033145
2016-03-22    0.033469
2016-03-15    0.033654
2016-03-09    0.033746
2016-03-19    0.033746
2016-03-29    0.033793
2016-03-10 

**Dates last seen**

In [38]:
# distribution of dates last seen
dates_last_seen = autos["last_seen"].str[:10]
dates_last_seen.value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001387
2016-03-06    0.004946
2016-03-07    0.006102
2016-03-08    0.008413
2016-03-09    0.010864
2016-03-10    0.011464
2016-03-11    0.014932
2016-03-12    0.027182
2016-03-13    0.010864
2016-03-14    0.012944
2016-03-15    0.017058
2016-03-16    0.018907
2016-03-17    0.031296
2016-03-18    0.008136
2016-03-19    0.016919
2016-03-20    0.021820
2016-03-21    0.022374
2016-03-22    0.023484
2016-03-23    0.019508
2016-03-24    0.021588
2016-03-25    0.019832
2016-03-26    0.018491
2016-03-27    0.016411
2016-03-28    0.021635
2016-03-29    0.022605
2016-03-30    0.025287
2016-03-31    0.025564
2016-04-01    0.024085
2016-04-02    0.027043
2016-04-03    0.025009
2016-04-04    0.025749
2016-04-05    0.117696
2016-04-06    0.202894
2016-04-07    0.117511
Name: last_seen, dtype: float64

In [39]:
# sort by percentage
dates_last_seen.value_counts(normalize=True, dropna=False).sort_values(ascending=True)

2016-03-05    0.001387
2016-03-06    0.004946
2016-03-07    0.006102
2016-03-18    0.008136
2016-03-08    0.008413
2016-03-09    0.010864
2016-03-13    0.010864
2016-03-10    0.011464
2016-03-14    0.012944
2016-03-11    0.014932
2016-03-27    0.016411
2016-03-19    0.016919
2016-03-15    0.017058
2016-03-26    0.018491
2016-03-16    0.018907
2016-03-23    0.019508
2016-03-25    0.019832
2016-03-24    0.021588
2016-03-28    0.021635
2016-03-20    0.021820
2016-03-21    0.022374
2016-03-29    0.022605
2016-03-22    0.023484
2016-04-01    0.024085
2016-04-03    0.025009
2016-03-30    0.025287
2016-03-31    0.025564
2016-04-04    0.025749
2016-04-02    0.027043
2016-03-12    0.027182
2016-03-17    0.031296
2016-04-07    0.117511
2016-04-05    0.117696
2016-04-06    0.202894
Name: last_seen, dtype: float64

In [39]:
# Understand the distribution of registration year
autos["registration_year"].describe()

count    21632.00000
mean      2002.52242
std         27.12203
min       1000.00000
25%       1999.00000
50%       2002.00000
75%       2005.00000
max       5000.00000
Name: registration_year, dtype: float64

The above data show:
* The minimum value is 1000, before car were invented
* The maximum value is 9999, many years into the future

### Dealing with Incorrect Registration Year Data

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitly inacurate

**Count cars fall outside the 1900-2016**

In [40]:
autos["registration_year"].value_counts().sort_index()

1000      1
1934      2
1937      1
1952      1
1956      2
       ... 
2018    300
2019      1
2800      1
4100      1
5000      1
Name: registration_year, Length: 69, dtype: int64

In [40]:
# Keep only entries with registeration year between 1910 and 2016
autos = autos[autos["registration_year"].between(1910,2016)]

In [41]:
autos["registration_year"].describe()

count    20617.000000
mean      2001.560654
std          5.600055
min       1934.000000
25%       1999.000000
50%       2002.000000
75%       2005.000000
max       2016.000000
Name: registration_year, dtype: float64

In [43]:
autos["registration_year"].value_counts(normalize=True).sort_index()

1934    0.000097
1937    0.000049
1952    0.000049
1956    0.000097
1958    0.000097
          ...   
2012    0.001310
2013    0.000582
2014    0.000485
2015    0.000243
2016    0.022942
Name: registration_year, Length: 62, dtype: float64

### Exploring Price by Brand

In [42]:
# the number of unique brands in total
autos["brand"].unique().shape

(40,)

In [43]:
# display distribution of brands
brands = autos["brand"].value_counts().sort_values(ascending=False).head(6).index

In [44]:
brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'ford', 'audi'], dtype='object')

In [45]:
autos.loc[autos["brand"] == "volkswagen", "price"]

7        1990
38       2850
40       3500
41       3200
57       4100
         ... 
49955    3400
49958    2900
49961    3200
49967    4999
49985    1000
Name: price, Length: 4402, dtype: int64

In [46]:
# aggregrate by brand

brand_mean_price = {}

for brand in brands:
    brand_autos = autos.loc[autos["brand"] == brand, "price"]
    price_autos = brand_autos.iloc[1]
    mean_price = price_autos.mean()
    brand_mean_price[brand] = mean_price
    
for brand in brand_mean_price:
    print(brand, brand_mean_price[brand])

volkswagen 2850.0
opel 2250.0
bmw 4700.0
mercedes_benz 2850.0
ford 1200.0
audi 3500.0


### Any relation between price and mileage

In [47]:
brand_mean_mileage = {}

for brand in brands:
    brand_autos = autos.loc[autos["brand"] == brand, "odometer_km"]
    mileage_autos = brand_autos.iloc[1]
    mean_mileage = mileage_autos.mean()
    brand_mean_mileage[brand] = mean_mileage

In [48]:
autos["odometer_km"]

0        150000
3         70000
4        150000
7        150000
13       150000
          ...  
49992    125000
49993    150000
49994    150000
49996    150000
49999    150000
Name: odometer_km, Length: 20617, dtype: int64

In [49]:
# convert the two dictionaries into pandas series
price_series = pd.Series(brand_mean_price)
mileage_series = pd.Series(brand_mean_mileage)

In [50]:
# Conver one Series into one column dataframe
df_price_mileage = pd.DataFrame(price_series, columns=['mean_price'])

In [51]:
# Add the second series as a column of the new dataframe
df_price_mileage["mean_mileage"] = mileage_series

In [52]:
# Print 
print(df_price_mileage)

               mean_price  mean_mileage
volkswagen         2850.0      125000.0
opel               2250.0      150000.0
bmw                4700.0      150000.0
mercedes_benz      2850.0      150000.0
ford               1200.0      150000.0
audi               3500.0      150000.0


In [53]:
df_price_mileage

mean_price  mean_mileage
volkswagen         2850.0      125000.0
opel               2250.0      150000.0
bmw                4700.0      150000.0
mercedes_benz      2850.0      150000.0
ford               1200.0      150000.0
audi               3500.0      150000.0

* Data cleaning next steps:
    * Identify categorical data that uses german words, translate them and map the values to their english counterparts
    * Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
    * See if there are particular keywords in the name column that you can extract as new columns
* Analysis next steps:
    * Find the most common brand/model combinations
    * Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the milage.
    * How much cheaper are cars with damage than their non-damaged counterparts?


In [ ]:
### Change german words to their english counterparts